File: llm_written_v1/gemini-2.5-pro/bfs/go/bfs_parallel.go
package llm_written

import (
	"runtime"
	"sort"
	"sync"
)

// PARALLEL_BFS_THRESHOLD is the number of vertices below which the sequential version is used.
const PARALLEL_BFS_THRESHOLD = 128

// Bfs is the public API that performs a parallel Breadth-First Search.
// It uses a level-synchronous approach, parallelizing the exploration of each level's frontier.
// It dispatches to a sequential implementation for small graphs to avoid overhead.
func Bfs(g Graph, startVertex int) []int {
	if len(g.Vertices) < PARALLEL_BFS_THRESHOLD {
		return BfsSequential(g, startVertex)
	}

	if _, exists := g.Vertices[startVertex]; !exists {
		return []int{}
	}

	// visited is a sync.Map to allow for concurrent, lock-free reads and writes.
	var visited sync.Map
	visited.Store(startVertex, true)

	result := []int{startVertex}
	frontier := []int{startVertex}

	for len(frontier) > 0 {
		numWorkers := runtime.NumCPU()
		if len(frontier) < numWorkers {
			numWorkers = len(frontier)
		}

		// Channel to collect newly discovered nodes from all workers for the next frontier.
		nextFrontierChan := make(chan []int, numWorkers)
		var wg sync.WaitGroup

		// Create a channel of tasks (nodes to visit in the current frontier).
		tasks := make(chan int, len(frontier))
		for _, node := range frontier {
			tasks <- node
		}
		close(tasks)

		// Start a bounded pool of workers.
		wg.Add(numWorkers)
		for i := 0; i < numWorkers; i++ {
			go func() {
				defer wg.Done()
				localNextFrontier := []int{}
				// Each worker pulls nodes from the tasks channel until it's empty.
				for node := range tasks {
					neighbors := g.Vertices[node]
					for _, neighbor := range neighbors {
						// LoadOrStore is an atomic operation.
						// If loaded is false, it means we successfully stored the key,
						// indicating this is the first time we've seen this node.
						if _, loaded := visited.LoadOrStore(neighbor, true); !loaded {
							localNextFrontier = append(localNextFrontier, neighbor)
						}
					}
				}
				nextFrontierChan <- localNextFrontier
			}()
		}

		// Wait for all workers to finish processing the current frontier.
		wg.Wait()
		close(nextFrontierChan)

		// Merge the results from all workers into a single slice for the next frontier.
		nextFrontier := []int{}
		for partialFrontier := range nextFrontierChan {
			nextFrontier = append(nextFrontier, partialFrontier...)
		}
		
		// If no new nodes were discovered, the traversal is complete.
		if len(nextFrontier) == 0 {
			break
		}

		// Crucially, sort the next frontier. This ensures that the traversal order
		// is deterministic across multiple runs, which is essential for testing and predictability.
		sort.Ints(nextFrontier)

		result = append(result, nextFrontier...)
		frontier = nextFrontier
	}

	return result
}

File: llm_written_v1/gemini-2.5-pro/bfs/cpp/bfs_parallel.cpp
#include "bfs_parallel.hpp"
#include "bfs_seq.hpp"
#include "graph.h"
#include <vector>
#include <queue>
#include <unordered_map>
#include <omp.h>
#include <algorithm>
#include <iostream>

// Threshold for switching to sequential implementation
const int PARALLEL_THRESHOLD = 0;

std::vector<int> bfs_parallel(Graph& g, int start_vertex) {
    if (g.vertices.size() < PARALLEL_THRESHOLD) {
        return bfs(g, start_vertex);
    }

    std::vector<int> result;
    if (g.vertices.count(start_vertex) == 0) {
        return result;
    }

    std::unordered_map<int, bool> visited;
    for(auto const& [key, val] : g.vertices) {
        visited[key] = false;
    }

    std::vector<int> frontier;
    
    visited[start_vertex] = true;
    frontier.push_back(start_vertex);

    while (!frontier.empty()) {
        for (int node : frontier) {
            result.push_back(node);
        }

        int num_threads = omp_get_max_threads();
        std::vector<std::vector<int>> local_next_frontiers(num_threads);

        #pragma omp parallel for schedule(dynamic)
        for (size_t i = 0; i < frontier.size(); ++i) {
            int current = frontier[i];
            int thread_id = omp_get_thread_num();
            
            // To ensure deterministic neighbor traversal, we sort them first.
            std::vector<int> neighbors = g.vertices.at(current);
            std::sort(neighbors.begin(), neighbors.end());

            for (int neighbor : neighbors) {
                bool is_visited = false;
                #pragma omp critical(visited_check)
                {
                    is_visited = visited[neighbor];
                }
                
                if (!is_visited) {
                    bool set_by_this_thread = false;
                    #pragma omp critical(visited_update)
                    {
                        if (!visited[neighbor]) {
                            visited[neighbor] = true;
                            set_by_this_thread = true;
                        }
                    }
                    if (set_by_this_thread) {
                        local_next_frontiers[thread_id].push_back(neighbor);
                    }
                }
            }
        }

        std::vector<int> next_frontier;
        for (int i = 0; i < num_threads; ++i) {
            next_frontier.insert(next_frontier.end(), local_next_frontiers[i].begin(), local_next_frontiers[i].end());
        }
        
        // Sort to ensure deterministic traversal order across runs
        std::sort(next_frontier.begin(), next_frontier.end());
        frontier = next_frontier;
    }

    return result;
}
